### PR TITLE
Network Security : Implement StrictOrigin and StrictOriginWhenCrossOr…

### PR DESCRIPTION
**Pull request by nmvk**
This pull request contains commit implementing initial steps for Improving Network Security project. As part of initial steps referer policy enums for strict-origin and strict-origin-when-cross-origin have been added to [hyper](https://github.com/hyperium/hyper/pull/943). Unit tests and additional logic has been added to handle these policies. Since enum changes are available on hyper version 0.9.11. We had to update hyper version to 0.9.11.

Hyper 0.9.11 depends on num_cpus 1.1.0. To avoid different version of num_cpus. We have updated rayon version from 0.4.0 to 0.4.3. Cargo.toml of util, style, geckolib, stylo component has been updated to use num_cpus version 1.1.0 instead of 0.2.2.  

---

- [ X] `./mach build -d` does not report any errors
- [ X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

…igin

Referer policy strict-origin and strict-origin-when-cross-origin changes have been implemented. Relevant unit test cases have been added. Enum for RefererPolicy has been added to hyper codebase and v 0.9.11 of hyper contains these changes.

This commit also contains changes related to upgrade of hyper from v0.9.10 to v0.9.11. Other dependencies changed are rayon, utils, num_cpus.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14059)

### CODE REVIEW

File: components/net/http_loader.rs [hunk 1]
```diff
@@ -437,6 +437,27 @@ fn no_referrer_when_downgrade_header(referrer_url: Url, url: Url) -> Option<Url>
     return strip_url(referrer_url, false);
 }
 
+/// https://w3c.github.io/webappsec-referrer-policy/#referrer-policy-strict-origin
+fn strict_origin(referrer_url: Url, url: Url) -> Option<Url> {
+    if referrer_url.scheme() == "https" && url.scheme() != "https" {
+        return None;
+    }
+    return strip_url(referrer_url, true);
```
Comment (nox) [inline components/net/http_loader.rs]: Nit: `return` is useless here.

File: components/net/http_loader.rs [hunk 2]
```diff
@@ -437,6 +437,27 @@ fn no_referrer_when_downgrade_header(referrer_url: Url, url: Url) -> Option<Url>
...
+/// https://w3c.github.io/webappsec-referrer-policy/#referrer-policy-strict-origin-when-cross-origin
+fn strict_origin_when_cross_origin(referrer_url: Url, url: Url) -> Option<Url> {
+    let cross_origin = referrer_url.origin() != url.origin();
+    if referrer_url.scheme() == "https" && url.scheme() != "https" {
+        return None;
+    } else {
+        if cross_origin {
+            return strip_url(referrer_url, true);
+        }
+        return strip_url(referrer_url, false);
```
Comment (nox) [inline components/net/http_loader.rs]: Nit: `return` is useless here and doing `strip_url(referrer_url, cross_origin)` is shorter.

File: components/net/http_loader.rs [hunk 3]
```diff
@@ -437,6 +437,27 @@ fn no_referrer_when_downgrade_header(referrer_url: Url, url: Url) -> Option<Url>
...
+fn strict_origin(referrer_url: Url, url: Url) -> Option<Url> {
+    if referrer_url.scheme() == "https" && url.scheme() != "https" {
+        return None;
+    }
+    return strip_url(referrer_url, true);
+}
+
+/// https://w3c.github.io/webappsec-referrer-policy/#referrer-policy-strict-origin-when-cross-origin
+fn strict_origin_when_cross_origin(referrer_url: Url, url: Url) -> Option<Url> {
+    let cross_origin = referrer_url.origin() != url.origin();
```
Comment (nox) [inline components/net/http_loader.rs]: Nit: you can do that after the `return None` shortcut.

File: components/net/http_loader.rs [hunk 4]
```diff
@@ -437,6 +437,27 @@ fn no_referrer_when_downgrade_header(referrer_url: Url, url: Url) -> Option<Url>
...
+    if referrer_url.scheme() == "https" && url.scheme() != "https" {
+        return None;
+    } else {
```
Comment (nox) [inline components/net/http_loader.rs]: Nit: no need for a `else` here, removing it reduces indentation which is nice.

**Review by nox — CHANGES_REQUESTED** [review]
Some nits.

File: components/net/http_loader.rs (hunk 1 above)
Comment (mrnayak) [inline components/net/http_loader.rs]: Can you please elaborate why return is not required?

Thanks,
Raghav

Comment (nox) [inline components/net/http_loader.rs]: It's the last expression of the function, and the last expression in a function is the function's return value.

Comment (mrnayak) [inline components/net/http_loader.rs]: Thank you,

I will change and commit the changes

**Comment by jdm** [comment]
Look at all those newly-passing tests! The timeout is #14067 and can be ignored, but the rest need to have their [expected results](https://github.com/servo/servo/blob/master/tests/wpt/README.md#updating-test-expectations) updated.

**Comment by jdm** [comment]
Also, it looks like there was a rebase error in cef/Cargo.lock: https://github.com/servo/servo/pull/14059/files#diff-f58e90618d14a4b6a043be78f4a4af21R2882

**Comment by gurudarshan266** [comment]
@jdm All the .ini files corresponding to PASSing testcases have been removed for strict-origin and strict-origin-when-cross-origin referrer policies

**Pull request merged**